posthog/tasks/process_scheduled_changes.py

```diff
@@ -0,0 +1,35 @@
+from posthog.models import ScheduledChange
+from django.utils import timezone
+from posthog.models import FeatureFlag
+from django.db import transaction, OperationalError
+
+models = {"FeatureFlag": FeatureFlag}
+
+
+def process_scheduled_changes() -> None:
+    try:
+        with transaction.atomic():
+            scheduled_changes = ScheduledChange.objects.select_for_update(nowait=True).filter(
```

Comment:
scalability question, as I don't know the django ORM that well: is that a generator streaming rows by chucks, or would we immediately pull all the rows that match, potentially killing the pod due to memory usage? Is there a way to make sure this is paginated somehow?

Reply:
So the QuerySet, once iterated, indeed fetches all the rows into memory. The solution is to use the `iterator()` method and specify the `chunk_size` - done just now.

I thought about the multi-hour outage case when we need to process a large backlog. Looking at the current flag volumes, and being very conservative, it’s hard to imagine having more than several tens, maybe hundreds of thousands of scheduled changes in the backlog after an outage. This should be doable in the order of 10s of seconds. The important point is that we are now processing the rows in batches, thus limiting the impact on memory.

Happy to discuss further @xvello; also adding @pauldambra to comment on the implementation.

Reply:
We have [this helper](https://github.com/PostHog/posthog/blob/d36e52c7f27d990b2f7bc5d5bc94a111069e1459/posthog/celery.py#L107-L125) so that we can add scheduled tasks that expire

Here, otherwise, an outage of 20 minutes would mean we have a queue of ten scheduled flag tasks to run

Celery will hammer through tasks as fast as it can, so I can imagine a race (although the changes look relatively safe to repeat)

So you could add these as expiring tasks but not I think a hard requirement here

Reply:
I'm actually already using `add_periodic_task_with_expiry`.
Perhaps not necessary since we only ever have 1 worker processing the task, but it doesn't hurt either.

Reply:
🤦 👍